File: backend/src/BMPCommerce.Domain/Entities/Cliente.cs
using System.Diagnostics.CodeAnalysis;
using System.Text.RegularExpressions;
using BMPCommerce.Domain.Common;

namespace BMPCommerce.Domain.Entities;

// Cliente é "CRUD honesto" (Doc 01 §6): sem value objects nem agregado —
// validações simples direto na entidade.
public class Cliente : BaseEntity
{
    private static readonly Regex EmailPattern = new(@"^[^@\s]+@[^@\s]+\.[^@\s]+$", RegexOptions.Compiled);

    public string Nome { get; private set; }

    public string? CpfCnpj { get; private set; }

    public string? Telefone { get; private set; }

    public string? Email { get; private set; }

    public string? Cidade { get; private set; }

    public string? Estado { get; private set; }

    public string? Observacoes { get; private set; }

    public bool Ativo { get; private set; }

    // Exigido pelo EF Core para materializar a entidade via reflexão.
    private Cliente()
    {
        Nome = null!;
    }

    public Cliente(
        string nome,
        string? cpfCnpj,
        string? telefone,
        string? email,
        string? cidade,
        string? estado,
        string? observacoes)
    {
        Preencher(nome, cpfCnpj, telefone, email, cidade, estado, observacoes);
        Ativo = true;
    }

    public void Atualizar(
        string nome,
        string? cpfCnpj,
        string? telefone,
        string? email,
        string? cidade,
        string? estado,
        string? observacoes,
        bool ativo)
    {
        Preencher(nome, cpfCnpj, telefone, email, cidade, estado, observacoes);
        Ativo = ativo;
        MarkAsUpdated();
    }

    [MemberNotNull(nameof(Nome))]
    private void Preencher(
        string nome,
        string? cpfCnpj,
        string? telefone,
        string? email,
        string? cidade,
        string? estado,
        string? observacoes)
    {
        if (string.IsNullOrWhiteSpace(nome))
        {
            throw new DomainException("Nome do cliente é obrigatório.");
        }

        Nome = nome.Trim();
        CpfCnpj = NormalizarCpfCnpj(cpfCnpj);
        Telefone = string.IsNullOrWhiteSpace(telefone) ? null : telefone.Trim();
        Email = NormalizarEmail(email);
        Cidade = string.IsNullOrWhiteSpace(cidade) ? null : cidade.Trim();
        Estado = NormalizarEstado(estado);
        Observacoes = string.IsNullOrWhiteSpace(observacoes) ? null : observacoes.Trim();
    }

    private static string? NormalizarCpfCnpj(string? valor)
    {
        if (string.IsNullOrWhiteSpace(valor))
        {
            return null;
        }

        var digitos = new string(valor.Where(char.IsDigit).ToArray());

        if (digitos.Length is not (11 or 14))
        {
            throw new DomainException("CPF/CNPJ inválido: informe 11 dígitos (CPF) ou 14 dígitos (CNPJ).");
        }

        return digitos;
    }

    private static string? NormalizarEmail(string? email)
    {
        if (string.IsNullOrWhiteSpace(email))
        {
            return null;
        }

        var normalizado = email.Trim().ToLowerInvariant();

        if (!EmailPattern.IsMatch(normalizado))
        {
            throw new DomainException($"Email '{email}' possui formato inválido.");
        }

        return normalizado;
    }

    private static string? NormalizarEstado(string? estado)
    {
        if (string.IsNullOrWhiteSpace(estado))
        {
            return null;
        }

        var normalizado = estado.Trim().ToUpperInvariant();

        if (normalizado.Length != 2)
        {
            throw new DomainException("Estado deve ser a sigla UF com 2 letras (ex: SP).");
        }

        return normalizado;
    }
}

File: backend/app/domain/cliente.py
"""Equivalente a Domain/Entities/Cliente.cs — "CRUD honesto" (Doc 01 §6): sem value
objects nem agregado, validações simples direto na entidade."""

from __future__ import annotations

import re

from app.core.exceptions import DomainException
from app.domain.common import Entity

_EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


class Cliente(Entity):
    def __init__(
        self,
        nome: str,
        cpf_cnpj: str | None,
        telefone: str | None,
        email: str | None,
        cidade: str | None,
        estado: str | None,
        observacoes: str | None,
    ) -> None:
        self._init_entity()
        self._preencher(nome, cpf_cnpj, telefone, email, cidade, estado, observacoes)
        self.ativo = True

    def atualizar(
        self,
        nome: str,
        cpf_cnpj: str | None,
        telefone: str | None,
        email: str | None,
        cidade: str | None,
        estado: str | None,
        observacoes: str | None,
        ativo: bool,
    ) -> None:
        self._preencher(nome, cpf_cnpj, telefone, email, cidade, estado, observacoes)
        self.ativo = ativo
        self.mark_as_updated()

    def _preencher(
        self,
        nome: str,
        cpf_cnpj: str | None,
        telefone: str | None,
        email: str | None,
        cidade: str | None,
        estado: str | None,
        observacoes: str | None,
    ) -> None:
        if not nome or not nome.strip():
            raise DomainException("Nome do cliente é obrigatório.")

        self.nome = nome.strip()
        self.cpf_cnpj = self._normalizar_cpf_cnpj(cpf_cnpj)
        self.telefone = telefone.strip() if telefone and telefone.strip() else None
        self.email = self._normalizar_email(email)
        self.cidade = cidade.strip() if cidade and cidade.strip() else None
        self.estado = self._normalizar_estado(estado)
        self.observacoes = observacoes.strip() if observacoes and observacoes.strip() else None

    @staticmethod
    def _normalizar_cpf_cnpj(valor: str | None) -> str | None:
        if not valor or not valor.strip():
            return None

        digitos = "".join(ch for ch in valor if ch.isdigit())

        if len(digitos) not in (11, 14):
            raise DomainException("CPF/CNPJ inválido: informe 11 dígitos (CPF) ou 14 dígitos (CNPJ).")

        return digitos

    @staticmethod
    def _normalizar_email(email: str | None) -> str | None:
        if not email or not email.strip():
            return None

        normalizado = email.strip().lower()

        if not _EMAIL_PATTERN.match(normalizado):
            raise DomainException(f"Email '{email}' possui formato inválido.")

        return normalizado

    @staticmethod
    def _normalizar_estado(estado: str | None) -> str | None:
        if not estado or not estado.strip():
            return None

        normalizado = estado.strip().upper()

        if len(normalizado) != 2:
            raise DomainException("Estado deve ser a sigla UF com 2 letras (ex: SP).")

        return normalizado

    @classmethod
    def from_persistence(cls, **kwargs) -> Cliente:
        obj = cls.__new__(cls)
        for key, value in kwargs.items():
            setattr(obj, key, value)
        return obj
